stacks-signer/src/runloop.rs

```diff
@@ -969,6 +1036,118 @@ impl<C: Coordinator> RunLoop<C> {
+                    .or_default();
+                // Broadcast the NACK message since it's the first time for this signer ID with the current coordinator metadata
+                if nack_sent_to.insert(packet_sender_signer_id) {
+                    warn!(
+                    "Local signer {} broadcasting a NACK message with its more recent coordinator metadata to remote signer {}",
+                    self.signing_round.signer_id, packet_sender_signer_id
+                );
+                    self.broadcast_nack_message(
+                        *current_signer_coordinator_metadata,
+                        packet_sender_signer_id,
+                    );
+                }
+            }
+        }
+    }
+
+    /// The current signer broadcasts a NACK when it has a more updated
+    /// chain view than the CoordinatorMetadata found in processed Packets
+    fn broadcast_nack_message(
+        &mut self,
+        current_signer_coordinator_metadata: CoordinatorMetadata,
+        target_signer_id: u32,
+    ) {
+        let nack_message = NackMessage::new(
+            self.signing_round.signer_id,
+            target_signer_id,
+            current_signer_coordinator_metadata,
+        );
+        if let Err(e) = self.stackerdb.send_message_with_retry(nack_message.into()) {
+            warn!(
+                "Failed to send NACK for outdated coordinator view to stacker-db: {:?}",
+                e
+            );
+        } else {
+            self.nack_messages_sent
+                .entry(current_signer_coordinator_metadata)
+                .or_default()
+                .insert(target_signer_id);
+        }
+    }
+
+    /// Processes broadcasted NACK message by adding it to the nack_messages_received map if the message is
+    /// targeted to the current signer and represents a more updated view of the Stacks chain
+    fn process_nack_message(
+        &mut self,
+        nack_message: NackMessage,
+        current_coordinator_metadata: CoordinatorMetadata,
+        stale_node_nack_policy: StaleNodeNackPolicy,
+    ) {
+        let remote_coordinator_metadata = nack_message.coordinator_metadata;
+        if nack_message.target_signer_id == self.signer_id
+            && remote_coordinator_metadata.stacks_block_height
+                > current_coordinator_metadata.stacks_block_height
+        {
+            self.nack_messages_received
+                .entry(remote_coordinator_metadata)
+                .or_default()
+                .insert(nack_message.sender_signer_id);
+
+            // Apply back-off delay if nack_threshold is reached
+            if let Some(threshold) = self.nack_threshold {
+                // Get the count of NACKs for the current coordinator metadata
+                let nack_count = self
+                    .nack_messages_received
+                    .get(&remote_coordinator_metadata)
+                    .map_or(0, |signers| signers.len());
+
+                if nack_count >= threshold as usize {
```

Comment:
Isn't `threshold` a percentage, not an absolute quantity?

Reply:
I accept it as percentage from the local signer in their local policy, but then calculate what that quantity would be based on total number of signers.